src/main/java/com/google/devtools/build/lib/sandbox/SymlinkedSandboxedSpawn.java

```diff
@@ -68,29 +76,122 @@ public SymlinkedSandboxedSpawn(
+            inputs,
+            inputsToCreate,
+            dirsToCreate,
+            sandboxExecRoot,
+            treeDeleter);
+        return;
+      }
+    }
+
+    if (SandboxStash.useInMemoryStashes()) {
+      Map<PathFragment, StashContents> stashContentsMap = new HashMap<>();
+      for (Map.Entry<PathFragment, Path> entry : inputs.getFiles().entrySet()) {
+        if (entry.getValue() == null) {
+          continue;
+        }
+        PathFragment parent = entry.getKey().getParentDirectory();
+        boolean parentWasPresent = !addParent(stashContentsMap, parent);
+        stashContentsMap
+            .get(parent)
+            .filesToPath()
+            .put(entry.getKey().getBaseName(), entry.getValue());
+        addAllParents(stashContentsMap, parentWasPresent, parent);
+      }
+      for (Map.Entry<PathFragment, PathFragment> entry : inputs.getSymlinks().entrySet()) {
+        if (entry.getValue() == null) {
+          continue;
+        }
+        PathFragment parent = entry.getKey().getParentDirectory();
+        boolean parentWasPresent = !addParent(stashContentsMap, parent);
+        stashContentsMap
+            .get(parent)
+            .symlinksToPathFragment()
+            .put(entry.getKey().getBaseName(), entry.getValue());
+        addAllParents(stashContentsMap, parentWasPresent, parent);
+      }
+
+      for (var outputDir :
+          Iterables.concat(
+              outputs.files().values().stream()
+                  .map(PathFragment::getParentDirectory)
+                  .collect(ImmutableList.toImmutableList()),
+              outputs.dirs().values())) {
+        PathFragment parent = outputDir;
+        boolean parentWasPresent = !addParent(stashContentsMap, parent);
+        addAllParents(stashContentsMap, parentWasPresent, parent);
+      }
+      StashContents main = StashContents.create();
+      main.dirEntries()
+          .put(SandboxStash.getWorkspaceName(), stashContentsMap.get(PathFragment.EMPTY_FRAGMENT));
+      SandboxStash.setPathContents(sandboxPath, main);
+    }
+  }
+
+  private static boolean addParent(
+      Map<PathFragment, StashContents> stashContentsMap, PathFragment parent) {
+    boolean parentWasPresent = true;
+    if (!stashContentsMap.containsKey(parent)) {
+      stashContentsMap.put(parent, StashContents.create());
+      parentWasPresent = false;
+    }
+    return !parentWasPresent;
+  }
+
+  private static void addAllParents(
+      Map<PathFragment, StashContents> stashContentsMap,
+      boolean parentWasPresent,
+      PathFragment parent) {
+    while (!parentWasPresent && parent.getParentDirectory() != null) {
+      PathFragment parentParent = parent.getParentDirectory();
+      if (stashContentsMap.containsKey(parentParent)) {
+        parentWasPresent = true;
+      } else {
+        stashContentsMap.put(parentParent, StashContents.create());
+      }
+      StashContents parentParentStashContents = stashContentsMap.get(parentParent);
```

Comment:
These lines perform between two and three map lookups. If you perform the `get` first, you can save one.